---
title: "1- Australia oropharingeal cancer preprocessing"
author: "Marc Elosua-Bayes"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  BiocStyle::html_document:
    toc: yes
    toc_float: yes
    number_sections: yes
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, out.width = "100%", fig.align='center', 
                      message = FALSE, warning = FALSE)
options(width = 1200)
```

## Introduction
In this Rmarkdown document we will load and preprocess samples 161429, 161430, 161431, and 161432.
These samples are human oropharyngeal cancer and were provided to use by Joseph Powell from the Garvand Institute of Medical Research.

Here is the methodology used to obtain and process this samples up to this point:

All patients provided informed consent for the collection of human specimens and data.  This was approved by the St Vincent’s Hospital Research Office (2019/PID04335) in accordance with the National Health and Medical Research Council’s National Statement of Ethical Conduct in Human Research.  Patients undergoing surgical resection for a locally advanced oropharyngeal cancer were recruited to the study.  After surgical removal, the anatomical pathologist dissected a sample of both the primary and nodal metastasis.   Samples were tumour banked in accordance with our ethically approved protocol.  

**Sample storage**
Within 30 minutes of collection, tumour samples were tumour banked.  Samples were cut into 1mm x 1mm chunks with a scalpel blade.    For Visium, a tissue chunk was snap frozen in OCT. After freezing, samples were moved to liquid nitrogen for long term storage.

Frozen tissue samples were processed using the Visium Spatial Gene Expression slide and reagent kit (10X Genomics, US) following the manufacturer’s instruction. Briefly, 10 μm sections were placed into the capture areas of the Visium slide. Tissue morphology was assessed with H&E staining and imaging using a Leica DM6000 microscope equipped with a 20x lens (Leica, DE). The imaged sections were then permeabilized for 12 minutes using the supplied reagents. The permeabilization condition was previously optimised using the Visium Spatial Tissue Optimisation slide and reagent kit (10X Genomics, US). After permeabilization, cDNA libraries were prepared, checked for quality and sequenced on a NovaSeq 6000 platform (Illumina, US). Around 300 million pair-ended reads were obtained for each tissue section. Read 1, i7 index and Read 2 were sequenced with 28, 8 and 98 cycles respectively. Reads were demultiplexed and mapped to the reference genome GRCh38 using the Space Ranger Software v1.0.0 (10X Genomics).


## Libraries
```{r}
library(Seurat)
library(dplyr)
library(cowplot)
source(here::here("misc/paths.R"))
source(here::here("utils/bin.R"))
source(here::here("utils/bin.r"))
```

## Create directory structure
```{r}
"{an_oro}/{robj_dir}" %>%
  glue::glue() %>%
  here::here() %>%
  dir.create(
    path = ,
    showWarnings = FALSE,
    recursive = TRUE)

"{an_oro}/{plt_dir}" %>%
  glue::glue() %>%
  here::here() %>%
  dir.create(
    path = ,
    showWarnings = FALSE,
    recursive = TRUE)
```

## Parameters
Set common parameters for the entire document.
```{r}
id_ls <- c("161429", "161430", "161431", "161432")
```

## Load data
We are going to start by loading the data
```{r}
st_ls <- lapply(id_ls, function(id) {
  print(id)
  tmp_se <- Seurat::Load10X_Spatial(data.dir = here::here(sprintf("data/australia_hn/%s/", id)),
                                  filename = sprintf("%s_filtered_feature_bc_matrix.h5", id),
                                  slice = id)
  return(tmp_se)
})
names(st_ls) <- id_ls
```

## QC
1st thing we do is remove empty gene
```{r}
for (id in names(st_ls)) {
  table(rowSums(as.matrix(st_ls[[id]]@assays$Spatial@counts)) == 0)
  
  keep_genes <- rowSums(as.matrix(st_ls[[id]]@assays$Spatial@counts)) != 0
  st_ls[[id]] <- st_ls[[id]][keep_genes, ]
}
```

Then we look at the number of reads, genes, and mitochondrial per spot to check how this tissue looks.
First we add the mitochondrial and ribosomal percentage within each spot.
```{r}
for (id in names(st_ls)) {
  
  # store mitochondrial percentage in object meta data
  st_ls[[id]] <- Seurat::PercentageFeatureSet(st_ls[[id]],
                                     pattern = "^MT-",
                                     col.name = "percent.mt")
  
  st_ls[[id]] <- Seurat::PercentageFeatureSet(st_ls[[id]],
                                     pattern = "^RPS|^RPS",
                                     col.name = "percent.rp")
}
```

### Visualize QC metrics
#### Descriptive histograms
```{r}
lapply(names(st_ls), function(id) {
  print(id)
  hist_qc <- QC_seurat_hist(se = st_ls[[id]],
                          assay = "Spatial",
                          slot = "counts",
                          nfeat = "nFeature_Spatial",
                          ncount = "nCount_Spatial",
                          pctmt = "percent.mt",
                          pctrp = "percent.rp")
  
  "{an_oro}/{plt_dir}/QC_histogram_{id}.pdf" %>%
    glue::glue() %>%
    here::here() %>%
    cowplot::save_plot(
      filename = .,
      plot = hist_qc,
      base_height = 9,
      base_width = 16)
  return(hist_qc)
})
```

When we look at all the histograms we appreciate there is a peak on 0 *genes per spot* and *total counts per spot*, these we need to look at in more detail since they could be empty spots which we would discard.

#### Spatial visualization
Next we want to visualize the spots on the tissue to assess their spatial behaviour. We are mainly interested to see where the empty spots map.
```{r}
lapply(names(st_ls), function(id) {
  print(id)
  sp_qc <- Seurat::SpatialFeaturePlot(object =  st_ls[[id]],
                                      features = c("nFeature_Spatial",
                                                   "nCount_Spatial",
                                                   "percent.mt",
                                                   "percent.rp"))
  
  "{an_oro}/{plt_dir}/QC_spatial_{id}.pdf" %>%
    glue::glue() %>%
    here::here() %>%
    cowplot::save_plot(
      filename = .,
      plot = sp_qc,
      base_height = 12,
      base_width = 12)

  return(sp_qc)
})
```

We see that on all tissues there are spots with 0 or near 0 counts per spot. We will remove them by setting a thresholds since these are low quality spots that would distort the analysis.

We will subset and remove those spots with < 500 counts.
```{r}
for (id in names(st_ls)) {
  print(id)
  st_ls[[id]] <- st_ls[[id]][, st_ls[[id]]$nCount_Spatial > 500]
}
```

Take a new look at the spatial plots to assess how we removed those spots
```{r}
lapply(names(st_ls), function(id) {
  print(id)
  sp_qc <- Seurat::SpatialFeaturePlot(object =  st_ls[[id]],
                                      features = c("nFeature_Spatial",
                                                   "nCount_Spatial",
                                                   "percent.mt",
                                                   "percent.rp"))
  
    "{an_oro}/{plt_dir}/QC_spatial_subset_{id}.pdf" %>%
    glue::glue() %>%
    here::here() %>%
    cowplot::save_plot(
      filename = .,
      plot = sp_qc,
      base_height = 12,
      base_width = 12)
  
  return(sp_qc)
})
```

161430 and 161431 appear to have regions with very low reads so we'll keep an eye on these 2 slices specially in the downstream analysis.

## Normalization
Once we've subset the data to remove low quality spots we will scale, normalize and cluster the data.
```{r}
for (id in names(st_ls)) {
  print(id)
  st_ls[[id]] <- Seurat::SCTransform(object = st_ls[[id]],
                                     assay = "Spatial")
  
  # Dimensionality reduction and clustering
  st_ls[[id]] <- Seurat::RunPCA(st_ls[[id]], assay = "SCT", verbose = TRUE)
  st_ls[[id]] <- Seurat::FindNeighbors(st_ls[[id]], reduction = "pca", dims = 1:30)
  st_ls[[id]] <- Seurat::FindClusters(st_ls[[id]], verbose = TRUE, resolution = c(0.09, 0.1, 0.25, 0.5, 0.75, 1))
  st_ls[[id]] <- Seurat::RunUMAP(st_ls[[id]], reduction = "pca", dims = 1:30)
}
```

Save processed dataset
```{r}
for (id in names(st_ls)) {
  "{an_oro}/{robj_dir}/processed_sp_oropharyngeal_{id}.RDS" %>%
    glue::glue() %>%
    here::here() %>%
    saveRDS(
      object = st_ls[[id]],
      file = .)
}
```

### Cluster visualization
Plot cluster on the spatial context at different resolutions.
```{r}
lapply(names(st_ls), function(id) {
  print(id)
  
  sp_clust <- Seurat::SpatialDimPlot(object =  st_ls[[id]],
                                      group.by = c("SCT_snn_res.0.1",
                                                   "SCT_snn_res.0.25")) +
    geom_point(color = NA)
  
  "{an_oro}/{plt_dir}/Cluster_spatial_{id}.pdf" %>%
    glue::glue() %>%
    here::here() %>%
    cowplot::save_plot(
      filename = .,
      plot = sp_clust,
      base_height = 9,
      base_width = 15)

  return(sp_clust)
})
```


## Session Info
```{r}
sessionInfo()
```
