---
author: "Marc Elosua-Bayes"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: yes
    toc_float: yes
    number_sections: yes
    df_print: paged
editor_options: 
  chunk_output_type: console
params: 
    sample_id: "Default!"
title: "`r sprintf('2-Australia Oropharyngeal Biological processing %s', {params$sample_id})`"
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
knitr::opts_knit$set(root.dir = "../..")
```

## Introduction
In this Rmarkdown document we are going to do some downstream analysis to try to elucidate what the clustered regions represent.
We are going to use the data obtained from script *1-australis_oroph_processing.Rmd*.

## Libraries
```{r}
library(Seurat)
library(dplyr)
library(ggplot2)
library(cowplot)
library(purrr)
library(DT)
library(org.Hs.eg.db)
library(GOstats)
library(progeny)

source(here::here("misc/paths.R"))
source(here::here("utils/bin.r"))
source(here::here("utils/bin.R"))
```

## Parameters
Set common parameters for the entire document.
```{r}
id_ls <- c("161429", "161430", "161431", "161432")
sample <- params$sample_id
sample <- "161429"
```

## Load data
We are going to start by loading the data
```{r}
sp_obj <- "{an_oro}/{robj_dir}/processed_sp_oropharyngeal_{sample}.RDS" %>%
  glue::glue() %>%
  here::here() %>%
  readRDS(file = .)
```

## Go enrichment
### Marker genes
Iterate over the spatial seurat objects, set Identities to the lowest resolution and save the marker genes.
```{r}
Seurat::Idents(sp_obj) <- sp_obj@meta.data[, "SCT_snn_res.0.1"]
markers_all <- Seurat::FindAllMarkers(object = sp_obj,
                                      assay = "SCT",
                                      slot = "data",
                                      only.pos = TRUE)

"{an_oro}/{robj_dir}/markers_{sample}_res.0.1.rds" %>%
  glue::glue() %>%
  here::here() %>%
  saveRDS(
    object = tmp_markers,
    file = .)
```

Interactive table to look at marker genes
```{r}
DT::datatable(data = markers_all,
              filter = "top",
              options = list(pageLength = 20, autoWidth = TRUE))
```


```{r echo = FALSE}
# Extracting the markers from each cluster
# Function to pull the gene variable
pull_col <- function(.x) {
  return(.x %>% pull(var = "gene"))
}

unique_markers <- lapply(marker_ls, function(marker_df) {
  markers_tmp_ls <- marker_df %>%
  dplyr::filter(p_val < 0.01) %>%
  dplyr::group_by(cluster) %>% 
  dplyr::group_split() %>% 
  purrr::map(.f = pull_col)

  # Next we find unique marker genes for each clusters
  unique_hn1_ls <- lapply(seq_len(length(markers_tmp_ls)), function(i) {
    markers_tmp_ls[[i]][! markers_tmp_ls[[i]] %in% unique(unlist(markers_tmp_ls[-i]))]
  })
  return(unique_hn1_ls)

})

names(unique_markers) <- names(marker_ls)
```

### Cluster annotation
#### GO analysis
For ALL the slices
```{r}
DE_df <- lapply(names(marker_ls), function(nm) {
  # print(nm)
  # Extract objects of interest
  markers_st <- marker_ls[[nm]]
  sp_obj <- st_ls[[nm]]
  
  tmp_df <- lapply(unique(markers_st$cluster), function(clust) {
    # Subset cluster of interest
    gene_de <- markers_st %>%
      dplyr::filter(cluster == all_of(clust)) %>%
      dplyr::pull(gene) %>%
      stringr::str_to_upper(.)
    
    go_clust <- gene_enrichment_GO(
      gene_de = gene_de,
      gene_universe = stringr::str_to_upper(rownames(sp_obj)))
    
    tmp <- summary(go_clust) %>%
      dplyr::filter(Size <= 300 & Size >=33 & Pvalue < 0.01 & Count >= 3) %>% 
      dplyr::mutate(
        sample = nm,
        cluster = clust)
    
    return(tmp)
  }) %>% 
  dplyr::bind_rows()
  
  return(tmp_df)
}) %>% dplyr::bind_rows()

# Save DE gene analysis
"{an_oro}/{robj_dir}/go_enrichment_all_res.0.1_{sample}.rds" %>%
  glue::glue() %>%
  here::here() %>%
  saveRDS(
    object = DE_df,
    file = .)
```

Load enrichment RDS files
```{r}
# Save DE gene analysis
DE_df <- "{an_oro}/{robj_dir}/go_enrichment_all_res.0.1_{sample}.rds" %>%
  glue::glue() %>%
  here::here() %>%
  readRDS(file = .)
```

##### Visualization
For all the slices
```{r}
plt_ls <- lapply(unique(as.character(tmp_de$cluster)), function(i) {
  # print(i)
  tmp_plt <- DE_df %>%
    dplyr::filter(cluster == i) %>%
    dplyr::arrange(desc(OddsRatio)) %>%
    head(25) %>% 
    ggplot(.) +
      geom_point(aes(x = OddsRatio,
                     y = reorder(Term, OddsRatio),
                     size = -Pvalue,
                     color = Pvalue)) +
      scale_color_gradient(low = "green",
                           high = "red") +
      labs(title = glue::glue("Sample: {sample}; Cluster: {i}"),
           y = "")
  # cowplot::save_plot(filename = sprintf("%s/%s/go_enrich_hn1_clust-%s.pdf",
  # an_aussie, plt_dir, i),
  #                    plot = tmp_plt)
  })

"{an_oro}/{plt_dir}/go_enrich_res.0.1_{sample}.pdf" %>%
  glue::glue() %>%
  here::here() %>%
  ggpubr::ggexport(
    plotlist = plt_ls, 
    filename = .,
    width = 10,
    height = 7)
```

### PROGENy
PROGENy (Pathway RespOnsive GENes) estimates the activity of relevant signaling pathways based on consensus gene signatures obtained from perturbation experiments, in other words, the footprint of the pathway on gene expression
We compute PROGENy pathway activity scores on the scRNA-seq data, and we then characterice the different clusters based on these scores.
```{r}
## We compute the Progeny activity scores and add them to our Seurat object
## as a new assay called Progeny. 
sp_obj <- progeny::progeny(
  sp_obj,
  scale = FALSE,
  organism = "Human",
  top = 500,
  perm = 1,
  assay_name = "Spatial",
  return_assay = TRUE)

## We can now directly apply Seurat functions in our Progeny scores. 
## For instance, we scale the pathway activity scores. 
sp_obj <- Seurat::ScaleData(sp_obj, assay = "progeny") 
```

Next we create a data frame with the specification of the spot id that belong to each cluster to match with the Progeny scores.
```{r}
## We transform Progeny scores into a data frame to better handling the results
progeny_scores_df <- 
    as.data.frame(t(GetAssayData(sp_obj, slot = "scale.data", 
        assay = "progeny"))) %>%
    tibble::rownames_to_column("Cell") %>%
    tidyr::gather(Pathway, Activity, -Cell)

Seurat::Idents(sp_obj) <- sp_obj@meta.data[, "SCT_snn_res.0.1"]
## We create a data frame with the specification of the spot id that belong to 
## each cluster to match with the Progeny scores.
CellsClusters <- data.frame(Cell = names(Idents(sp_obj)), 
    CellType = as.character(Idents(sp_obj)),
    stringsAsFactors = FALSE)

## We match Progeny scores with the cell clusters.
progeny_scores_df <- inner_join(progeny_scores_df, CellsClusters, by = "Cell")

## We summarize the Progeny scores by cellpopulation
summarized_progeny_scores <- progeny_scores_df %>% 
    group_by(Pathway, CellType) %>%
    summarise(avg = mean(Activity), std = sd(Activity))
```

Lastly we want to visualize the pathways scores in a heatmap
```{r}
## We prepare the data for the plot
summarized_progeny_scores_df <- summarized_progeny_scores %>%
    dplyr::select(-std) %>%   
    tidyr::spread(Pathway, avg) %>%
    data.frame(row.names = 1, check.names = FALSE, stringsAsFactors = FALSE) 

paletteLength = 100
myColor = colorRampPalette(c("Darkblue", "white","red"))(paletteLength)

progenyBreaks <- c(seq(min(summarized_progeny_scores_df), 0, 
                      length.out = ceiling(paletteLength / 2) + 1),
                  seq(max(summarized_progeny_scores_df) / paletteLength, 
                      max(summarized_progeny_scores_df), 
                      length.out = floor(paletteLength / 2)))

"{an_oro}/{plt_dir}/progeny_hmap_{sample}.png" %>%
  glue::glue() %>%
  here::here() %>%
  pheatmap::pheatmap(
    mat = t(summarized_progeny_scores_df[, -1]),
    fontsize = 14,
    fontsize_row = 10,
    color = myColor,
    breaks = progenyBreaks,
    main = "PROGENy (500)",
    angle_col = 45,
    treeheight_col = 0,
    border_color = NA,
    filename = .)

progeny_hmap <- pheatmap::pheatmap(mat = t(summarized_progeny_scores_df[, -1]),
                                 fontsize = 14,
                                 fontsize_row = 10,
                                 color = myColor,
                                 breaks = progenyBreaks,
                                 main = "PROGENy (500)",
                                 angle_col = 45,
                                 treeheight_col = 0,
                                 border_color = NA)
```

## Session Info
```{r}
sessionInfo()
```

